Updates Android builder to be more encapsulated, configurable and complete

rharter opened the pull request:
This is a fairly substantial change to the Android builder, so I'd welcome discussion.

The changes to the Android `Dockerfile` and sample `cloudbuild.yaml` included here are based on [my prototype](https://github.com/pixiteapps/android-cloud-build).

There are a couple main points that I think this PR addresses that will make integrating Google Cloud Build into an Android workflow much easier for new users.

## 1. Android images should be flexible.

The original image configuration was cumbersome and highly customized, instead of meeting the needs of most users and allowing them to optimize from there.

With this change user's can easily (with 1 argument) configure the generated Android image based on their target API.  This is the most common thing that Android developers need to change, since other missing packages will be downloaded on demand.

From here it's easy to optimize the build by adding dependent packages, but user's don't need to do that from the start.

For users needing the NDK for their builds, the `cloudbuild-ndk.yaml` takes care of using the appropriate Dockerfile to include that.

## 2. Build settings live in the build config

Originally the build cache caching was in two places, the docker image (via the `gradle-build` command), and the `cloudbuild.yaml` configuration file (via the `gsutil` download at the right path).

This is cumbersome, as changes to the caching require both a new image and updates to the build script, but also require the build script to use commands that are different from the commands users run on their local machines.

It also meant that the build config would be inflated and compressed for every build step that ran gradle.

By removing the `gradle-build` script, and truly making caching part of the build configuration via the `cloudbuild.yaml` file, all of these problems are solved, allowing the images to be used almost without modification.

rharter commented on android/examples/build-test-deploy.yaml:
```diff
@@ -0,0 +1,232 @@
+steps:
+
+#
+# Get a build number
```
While this section isn't imperative for Android builds, incrementing build numbers is something that other major CI solutions offer and are used for many Android builds (for things like APK version codes).  This allows this sample to be used with extremely little modification by developers to try out Cloud Build and get started quickly.

nof20 replied:
Kicking the tires & getting started quickly sounds like an important goal, especially since mobile developers might be less familiar with containers in general.

Is it possible to add a simpler example which doesn't require setup of e.g. GCS buckets, Cloud KMS keys, etc.?  In other places we download and build sample code used in a quick start tutorial.

rharter replied:
Yeah, I like that idea. I'll work on getting a simpler sample together.

rharter commented on android/examples/build-test-deploy.yaml:
```diff
@@ -0,0 +1,232 @@
+    path: '/config'
+
+#
+# Extract the cache
```
Instead of having cache configuration live both in the `cloudbuild.yaml` **and** the `gradle-build` file, this section moves it to the config file, making the docker images far more flexible.

nof20 replied:
Sounds great.

johnlabarge replied:
This is very cool.

rharter commented on android/examples/build-test-deploy.yaml:
```diff
@@ -0,0 +1,232 @@
+    path: '/build_cache'
+
+#
+# Decrypt secret files
```
Another thing that's not imperative for Android but is an extremely common use case as Android builds require things like keystores in order to deploy to Play.

nof20 replied:
Agreed

rharter commented on android/examples/build-test-deploy.yaml:
```diff
@@ -0,0 +1,232 @@
+  - '-c'
+  - |
+    source build_environment.sh
+    [[ "$BRANCH_NAME" == "master" ]] && ./gradlew --stacktrace -Dtrace :app:publishReleaseApk || echo "Skipping Play Store distribution..."
```
Would love to see either configurable workflows ala CircleCI, or at least simple branch filtering to avoid this type of thing.

nof20 replied:
Understood.  You can set up [different builds to run from different branches](https://cloud.google.com/cloud-build/docs/running-builds/automate-builds) in Cloud Build, but can't introspect the branch the build is running from.

rharter commented on tar/README.md:
```diff
@@ -0,0 +1,28 @@
+# tar
```
This mirrors the "command images" found in the official `cloud-builders` repo.

nof20 replied:
Sounds good

nof20 commented on android/cloudbuild.yaml:
```diff
@@ -1,24 +1,27 @@
-  - |
-    gsutil rsync -d -r "${_LICENSES_BUCKET}" licenses || exit 0
-  id: copy-licenses
+  
```
Nit: remove trailing whitespace.
```suggestion

```

nof20 commented on android/cloudbuild.yaml:
```diff
@@ -1,24 +1,27 @@
-images: ['gcr.io/$PROJECT_ID/android-builder']
+  - 'gcr.io/$PROJECT_ID/android:base'
+
+images: 
```
Nit: remove trailing whitespace.
```suggestion
images:
```

nof20 commented on android/README.md:
```diff
@@ -1,109 +1,116 @@
+
+Create a `cloudbuild.yaml` file based on your needs and save it in the root of your project.
+
+Check out the [complete sample](samples/build-test-deploy.yaml) for some ideas of how to accomplish this.
```
s/sample/example/
```suggestion
Check out the [complete example](examples/build-test-deploy.yaml) for some ideas of how to accomplish this.
```

nof20 commented on tar/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM launcher.gcr.io/google/ubuntu16_04
```
What about using the Alpine base image instead?  It has minimal packages installed, so loads much faster and has a far smaller attack surface.  It's also cached on Cloud Build workers, and reduces build time from 35 seconds to 14 seconds.
```suggestion
FROM alpine
```

nof20 commented on tar/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM launcher.gcr.io/google/ubuntu16_04
+
+RUN apt-get update && \
```
```suggestion
RUN apk update && \
```

nof20 commented on tar/Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM launcher.gcr.io/google/ubuntu16_04
+
+RUN apt-get update && \
+    apt-get -y install tar bzip2 gzip
```
```suggestion
    apk install tar bzip2 gzip
```

nof20 commented on android/README.md:
```suggestion
tagged with `google-cloud-build`.
```

nof20 commented on android/Dockerfile-sdk:
```diff
@@ -0,0 +1,8 @@
+# we use a gcr.io image and not openjdk:8-jdk-slim because it loads faster in the google Google Cloud Build environment
```
Nit: suggest you remove this line and keep only the copy in `Dockerfile-base`, where it's in the right context.  Anyone building using this Dockerfile would have to pull `gcr.io/$project_id/android:base` from Container Registry.

rharter commented:
I've updated this as requested.  Please let me know if there's anything else needed, @nof20.

johnlabarge commented:
@rharter  Thanks so much for the contributions!

rharter commented:
@johnlabarge absolutely!  

I have run into issues with the alpine image as I've exercised this some more (specifically for Android), so I'd hold off a little while I work that out.

I'm happy to close this and reopen when that's solved, or I can leave this "on hold".

nof20 commented:
@rharter Thanks for all these changes.  If you're having compatibility problems with Alpine in the Android images, fall back to gcr.io/cloud-builders/javac.  (My suggestion to use Alpine was just for the tar image because it's so simple - sorry if that was misleading.)

CLA is now fully approved too.  I'm good to approve this as it stands.

rharter commented:
Thanks, @nof20. I've reverted the Android image to gcr.io/cloud-builders/javac and added some scripts to the image to help with Android builds.  I'm currently using this image and script in my project, so can verify that it is working well.

nof20 commented:
Thanks!
